File: backend/utils/mpesaService.js
const axios = require('axios');

class MpesaService {
  constructor() {
    this.consumerKey = process.env.MPESA_CONSUMER_KEY;
    this.consumerSecret = process.env.MPESA_CONSUMER_SECRET;
    this.shortCode = process.env.MPESA_SHORTCODE;
    this.passKey = process.env.MPESA_PASSKEY;
    this.baseUrl = 'https://sandbox.safaricom.co.ke';
  }

  async generateAccessToken() {
    const auth = Buffer.from(`${this.consumerKey}:${this.consumerSecret}`).toString('base64');
    try {
      const response = await axios.get(`${this.baseUrl}/oauth/v1/generate?grant_type=client_credentials`, {
        headers: {
          'Authorization': `Basic ${auth}`
        }
      });
      return response.data.access_token;
    } catch (error) {
      console.error('Error generating M-Pesa access token:', error);
      throw error;
    }
  }

  async initiateSTKPush(phoneNumber, amount, accountReference) {
    try {
      const accessToken = await this.generateAccessToken();
      const timestamp = new Date().toISOString().replace(/[:-]/g, '').split('.')[0];
      const password = Buffer.from(`${this.shortCode}${this.passKey}${timestamp}`).toString('base64');

      const payload = {
        BusinessShortCode: this.shortCode,
        Password: password,
        Timestamp: timestamp,
        TransactionType: 'CustomerPayBillOnline',
        Amount: amount,
        PartyA: phoneNumber,
        PartyB: this.shortCode,
        PhoneNumber: phoneNumber,
        CallBackURL: process.env.CALLBACK_URL,
        AccountReference: accountReference,
        TransactionDesc: 'SMARTSTAY CHUKA Subscription Payment'
      };

      const response = await axios.post(
        `${this.baseUrl}/mpesa/stkpush/v1/processrequest`,
        payload,
        {
          headers: {
            'Authorization': `Bearer ${accessToken}`,
            'Content-Type': 'application/json'
          }
        }
      );

      return response.data;
    } catch (error) {
      console.error('Error initiating M-Pesa STK push:', error);
      throw error;
    }
  }

  async validateCallback(data) {
    try {
      if (data.Body.stkCallback.ResultCode === 0) {
        return {
          success: true,
          data: data.Body.stkCallback.CallbackMetadata.Item
        };
      } else {
        return {
          success: false,
          message: 'Payment failed'
        };
      }
    } catch (error) {
      console.error('Error validating M-Pesa callback:', error);
      throw error;
    }
  }
}

module.exports = new MpesaService();

File: mobile/lib/services/mpesa_service.dart
import 'package:http/http.dart' as http;
import 'dart:convert';
import 'dart:async';

class MpesaService {
  static const String _baseUrl = 'https://sandbox.safaricom.co.ke';
  static const String _consumerKey = 'YOUR_CONSUMER_KEY';
  static const String _consumerSecret = 'YOUR_CONSUMER_SECRET';
  static const String _shortCode = '174379'; // Sandbox shortcode
  static const String _passKey =
      'bfb279f9aa9bdbcf158e97dd1a503b30015eee520bd5d0925ba8d7d13b0afdbf';

  // Get OAuth token
  static Future<String?> _getAccessToken() async {
    try {
      final auth =
          'Basic ${base64Encode(utf8.encode('$_consumerKey:$_consumerSecret'))}';
      final response = await http.get(
        Uri.parse('$_baseUrl/oauth/v1/generate?grant_type=client_credentials'),
        headers: {'Authorization': auth},
      );

      if (response.statusCode == 200) {
        final body = jsonDecode(response.body);
        return body['access_token'];
      }
      return null;
    } catch (e) {
      print('Token error: $e');
      return null;
    }
  }

  // Initiate STK Push
  static Future<Map<String, dynamic>?> initiateSTKPush({
    required String phoneNumber,
    required double amount,
    required String accountRef,
  }) async {
    try {
      final token = await _getAccessToken();
      if (token == null) return null;

      final timestamp =
          DateTime.now().toString().replaceAll(RegExp(r'[^\d]'), '').substring(0, 14);

      final password = base64Encode(utf8.encode(
          '$_shortCode$_passKey$timestamp'));

      final body = {
        'BusinessShortCode': _shortCode,
        'Password': password,
        'Timestamp': timestamp,
        'TransactionType': 'CustomerPayBillOnline',
        'Amount': amount.toInt(),
        'PartyA': phoneNumber,
        'PartyB': _shortCode,
        'PhoneNumber': phoneNumber,
        'CallBackURL':
            'https://your-backend.com/api/payments/mpesa-callback',
        'AccountReference': accountRef,
        'TransactionDesc': 'Chuka Hostels Subscription',
      };

      final response = await http.post(
        Uri.parse('$_baseUrl/mpesa/stkpush/v1/processrequest'),
        headers: {
          'Authorization': 'Bearer $token',
          'Content-Type': 'application/json',
        },
        body: jsonEncode(body),
      );

      if (response.statusCode == 200) {
        return jsonDecode(response.body);
      }
      return null;
    } catch (e) {
      print('STK Push error: $e');
      return null;
    }
  }

  // Check transaction status
  static Future<Map<String, dynamic>?> checkTransactionStatus({
    required String checkoutRequestId,
  }) async {
    try {
      final token = await _getAccessToken();
      if (token == null) return null;

      final timestamp =
          DateTime.now().toString().replaceAll(RegExp(r'[^\d]'), '').substring(0, 14);

      final password = base64Encode(utf8.encode(
          '$_shortCode$_passKey$timestamp'));

      final body = {
        'BusinessShortCode': _shortCode,
        'Password': password,
        'Timestamp': timestamp,
        'CheckoutRequestID': checkoutRequestId,
      };

      final response = await http.post(
        Uri.parse(
            '$_baseUrl/mpesa/stkpushquery/v1/query'),
        headers: {
          'Authorization': 'Bearer $token',
          'Content-Type': 'application/json',
        },
        body: jsonEncode(body),
      );

      if (response.statusCode == 200) {
        return jsonDecode(response.body);
      }
      return null;
    } catch (e) {
      print('Status check error: $e');
      return null;
    }
  }
}
